torch/utils/data/datapipes/datapipe.py

```diff
@@ -93,11 +94,10 @@ def class_function(cls, enable_df_api_tracing, source_dp, *args, **kwargs):
     def __getstate__(self):
         if IterDataPipe.getstate_hook is not None:
             return IterDataPipe.getstate_hook(self)
-        # TODO: Fix `dill` circular dependency - https://github.com/pytorch/data/issues/237
         if DILL_AVAILABLE:
             state_dict = {}
             for k, v in self.__dict__.items():
-                if callable(v):
+                if isfunction(v) and not ismethod(v):
```

Comment:
By not applying the special serialization to methods, the example given @pmeier is not an issue. In general, I think it is fair to assume that methods written in a `IterDataPipe` should be serializable by `pickle`, otherwise a custom `__getstate__` should be provided in the class as well.

I don't think any method directly uses `lambda` so I believe this should work. Any edge cases that you can think of?

Reply:
How about the method not bounded to `IterDataPipe`?
If the method is bounded into a function, I am not sure if this is possible to pickle. If not, we have to try using dill.

Reply:
I think the issue only happens for methods bound to `IterDataPipe`'s, right? Maybe we can be strict here to avoid other issues.

Reply:
I believe that with this condition `isfunction(v) and not ismethod(v)` here, only non-method function within a `IterDataPipe` will go through the custom serialization function `serialize_fn`.

This should not impact any method or other `IterDataPipe`. Do you have an example where this may not work?

Reply:
@ejguan @pmeier @VitalyFedyunin Checking if people have any thoughts on this

Reply:
Let's list all use cases of this function:
1. traverse with `only_datapipe=True`
2. traverse with `only_datapipe=False`
3. multiprocessing

For scenario 2 and 3, we need to pickle/dill functions or methods. Then, I have a noob question. I believe both `pickle` and `dill` are able to handle circular reference by themselves, could we simply using `dill` when it's available during `traverse` and `multiprocessing`.

Not sure if it's possible for `multiprocessing` though.

Reply:
Essentially, you are proposing a backend switch (i.e. always use `dill` when available, and don't return a tuple of serialized object and serialization type)?

I can see that working for scenario 2, but it seems like `multiprocessing` will still use `pickle` unless `pickle.Pickler` is monkey-patched (seems too hacky). But I'm also a noob on this, anyone else have a better suggestion?

I would think by editing the `__getstate__` and `__setstate__` of the class as we do here, `dill` should be properly used in the right situation when available.

Reply:
I just updated the logic to check if `fn` is a lambda rather than trying to `pickle.dumps` `fn` first.

I also added this test case that has a lambda attribute (and uses it) and the test passes without any recursive error as long as `dill` is installed`.
```python
class LambdaIterDataPipe(IterDataPipe):
        def add_one(self, x):
            return x + 1

        def classify(self, x):
            return 0

        def __init__(self, fn, source_dp=None):
            self.fn = fn
            self.lambda_fn = lambda x: x + 1
            self.source_dp = source_dp if source_dp else dp.iter.IterableWrapper([1, 2, 4])
            self._dp = self.source_dp.map(self.add_one).map(self.lambda_fn).demux(2, self.classify)[0]

        def __iter__(self):
            yield from self._dp
```
```python
self.assertTrue(list(TestCircularSerialization.LambdaIterDataPipe(lambda x: x + 1)) ==
                list(dill.loads(dill.dumps(TestCircularSerialization.LambdaIterDataPipe(lambda x: x + 1)))))
_ = traverse(TestCircularSerialization.LambdaIterDataPipe(fn=_fake_fn), only_datapipe=True)
_ = traverse(TestCircularSerialization.LambdaIterDataPipe(fn=_fake_fn), only_datapipe=False)
```

Reply:
Just want to double check if it works when the function is embedded into a nested data structure?
```
self.lambda_fn = [lambda x: x + 1, ]
```
This lambda function itself won't be handled by the special serialization. Am I right?

Reply:
Yes, that is correct. But if `dill.dumps` is used, it will still work fine.

I think it is worth re-stating the scope of the problem that we are attempting to solve here.

At first, we were only serializing certain function arguments of built-in DataPipes (e.g. `fn` that get passed into `Mapper`), as you can see the code on the left-hand side in [this PR](https://github.com/pytorch/pytorch/pull/72896/files). When `dill` is available, we serialize/deserialize with `dill` for those `fn` within built-in DataPipe.

In this PR, we are also targeting something similar - i.e. non-method functions within DataPipes. I think that is a reasonable coverage since it covers all the possible lambdas that can exist within the built-in DataPipes. (I'm going to clean up this PR slightly and add a comment to make this intention more explicitly).

If the user chooses to do something more unusual, I think they should either 1) write their own `__getstate__` and `__setstate__` in their custom `IterDataPipe` (as we do for `demux` and `fork` to avoid storing the buffer) or 2) use `dill.dumps`.

Reply:
That said, I think it will be useful to have a backend switch within DataLoader2 (specifically allowing users to choose to serialize with `pickle` or `dill`). Thoughts?

Reply:
I think the reason we bring the serialization logic from each DataPipe instance to this base class is for simplicity and generalization. There are too many corner cases we didn't consider at the beginning. If we want to let users to do custom `__getstate__` and `__setstate__`, why not we move the special serialization logic to the required `DataPipe` like `Mapper`, `Filter`, etc. And, let them to handle `lambda` using `dill` if available.

> That said, I think it will be useful to have a backend switch within DataLoader2 (specifically allowing users to choose to serialize with `pickle` or `dill`). Thoughts?

This is reasonable.

Reply:
What other common corner cases do you think we should handle here assuming we will add a backend switch within DataLoader2? A `list` of unpicklable objects?

Reply:
For example of the `DataPipe` with `buffer`, we would never know the element type within the `buffer` when we want to do serialization/checkpoint in the middle of epoch.

Reply:
Right, and current that is handled by specific DataPipes' `__getstate__` and `__setstate__` ([for example](https://github.com/pytorch/pytorch/blob/gh/nivekt/45/head/torch/utils/data/datapipes/iter/combining.py#L351-L380)). IIUC you are suggesting that we should handle that here within this base class as well?

My view is that the users should 1) write their own` __getstate__` and `__setstate__` in their custom `IterDataPipe`  or 2) use `dill.dumps`.

Reply:
The demux example has a problem, we are doing hack way for serialization. We would lose all the information inside the buffer as we change it to empty deque during serialization.

If we are focusing on the build-in DataPipe we provided, we will try our best to make them pickable. But, for the objects that user specified within these build-in DataPipe. as we have zero control, I think we would better to serialize by pickle first. And, if it doesn't work, we could fall back to `dill` when available or `DL2` specifies the serialization method.

Reply:
Let me try something else.

Reply:
For the `demux` example, I agree that the buffer will lose information. That is what I plan to address with snapshotting - the buffer's state should be restored when you restore/load a snapshot.

The various ways that I have tried to use `pickle.dumps` or `dill.dumps` within `__getstate__` are causing infinite recursion. I just tried a few different ways today but no luck. I am open to suggestion. I think it will be better for `DL2` to specify the serialization method rather than doing it within `__getstate__`.

Reply:
Based on the use case of serialization, I don't think we need to implement `__getstate__` or `__setstate__`.
1. traverse with `only_datapipe=True`: no need to worry about the object 
2. traverse with `only_datapipe=False`: We could always use `dill` if available in this case 
3. multiprocessing or serialization. We could let DL2 to switch serialization method to `pickle.dumps` or `dill.dumps` on DataPipe rather than calling them within `__getstate__`.

Reply:
cc: @VitalyFedyunin